tests/php/Extension/Publishable/PublishableSiteTreeTest.php

```diff
@@ -0,0 +1,330 @@
+    }
+
+    public function testObjectsActionUnpublishRecursiveInclusion(): void
+    {
+        // Test that recursive parents and children
+        SiteTree::config()->set('regenerate_parents', PublishableSiteTree::RELATION_INCLUDE_RECURSIVE);
+        // Regardless of configuration, we expect children to be added in objectsToDelete() on unpublish action
+        SiteTree::config()->set('regenerate_children', PublishableSiteTree::RELATION_INCLUDE_RECURSIVE);
+
+        $page = $this->objFromFixture(SiteTree::class, 'page3');
+        $childPage = $this->objFromFixture(SiteTree::class, 'page4');
+        $grandchildPage = $this->objFromFixture(SiteTree::class, 'page5');
+        $virtualPage = $this->objFromFixture(VirtualPage::class, 'page1');
+        $parentPage = $this->objFromFixture(SiteTree::class, 'page2');
+        $grandparentPage = $this->objFromFixture(SiteTree::class, 'page1');
+
+        // We expect the following pages to be deleted
+        $expectedDeleteIds = [
+            $page->ID,
+            $childPage->ID,
+            $grandchildPage->ID,
+            $virtualPage->ID,
+        ];
+
+        // We expect the following pages to be updated
+        $expectedUpdateIds = [
+            $parentPage->ID,
+            $grandparentPage->ID,
+        ];
+
+        $context = [
+            'action' => 'unpublish',
+        ];
+
+        $toDelete = $page->objectsToDelete($context);
+        $toUpdate = $page->objectsToUpdate($context);
+
+        // There should be the same 4 pages being deleted
+        $this->assertCount(4, $toDelete);
+        // There should be 2 object being updated (our parent and grandparent pages)
+        $this->assertCount(2, $toUpdate);
+
+        // Check that the expected Pages are represented in our objectsToDelete() (order doesn't matter)
+        $this->assertEqualsCanonicalizing($expectedDeleteIds, $this->getIdsFromArray($toDelete));
+        // Check that the expected Pages are represented in our objectsToUpdate() (order doesn't matter)
+        $this->assertEqualsCanonicalizing($expectedUpdateIds, $this->getIdsFromArray($toUpdate));
+    }
+
+    public function testUrlsToCache()
+    {
+        // Page class is required because RedirectorPage extends Page
+        if (!class_exists(Page::class)) {
+            $this->markTestSkipped('This unit test requires the Page class');
+        }
+        $page = new Page(['Title' => 'MyPage']);
+        $id = $page->write();
+        $this->assertSame(['http://example.com/mypage' => 0], $page->urlsToCache());
+        $redirectorPage = new RedirectorPage(['Title' => 'MyRedirectorPage']);
+        $redirectorPage->LinkToID = $id;
+        $redirectorPage->write();
+        $this->assertSame(['http://example.com/myredirectorpage' => 0], $redirectorPage->urlsToCache());
+    }
+
+    /**
+     * @param SiteTree[] $pages
+     */
+    protected function getIdsFromArray(array $pages): array
+    {
+        $ids = [];
+
+        foreach ($pages as $page) {
```

Comment:
[nitpick] Can use `array_map` instead.

Reply:
IMO `array_map` isn't as visually clear, and it adds an extra (hard to identify) item in stack traces.